File: bot/discord_bot.py
import discord
from discord.ext import commands
import aiohttp
import asyncio
import os
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

# Bot configuration
BOT_TOKEN = os.getenv('DISCORD_BOT_TOKEN')
CLIENT_ID = os.getenv('DISCORD_CLIENT_ID', '1397971356490006558')
VERIFIED_ROLE_ID = os.getenv('DISCORD_VERIFIED_ROLE_ID', '1397091951504916531')
API_BASE = 'https://7tk0rize--admin-api.functions.blink.new'
VERIFICATION_URL = 'https://discord-verification-bot-webpage-7tk0rize.sites.blink.new'

# Bot setup
intents = discord.Intents.default()
intents.message_content = True
intents.guilds = True
intents.members = True

bot = commands.Bot(command_prefix='!', intents=intents)

@bot.event
async def on_ready():
    print(f'✅ {bot.user} is now online!')
    print(f'📊 Connected to {len(bot.guilds)} servers')

def is_admin():
    """Check if user has administrator permissions"""
    def predicate(ctx):
        return ctx.author.guild_permissions.administrator
    return commands.check(predicate)

@bot.command(name='setup_verify')
@is_admin()
async def setup_verify(ctx):
    """Set up verification for the server"""
    try:
        # Create verification URL with server info
        verify_url = f"{VERIFICATION_URL}?guild_id={ctx.guild.id}&guild_name={ctx.guild.name}"
        
        embed = discord.Embed(
            title="🔐 Server Verification Setup",
            description="Click the button below to verify and gain access to this server!",
            color=0x5865F2
        )
        embed.add_field(
            name="How it works:",
            value="1. Click 'Verify with Discord'\n2. Authorize the application\n3. Get the verified role automatically",
            inline=False
        )
        embed.add_field(
            name="Verification URL:",
            value=f"[Click here to verify]({verify_url})",
            inline=False
        )
        embed.set_footer(text=f"Server ID: {ctx.guild.id}")
        
        # Create view with button
        view = VerificationView(verify_url)
        
        await ctx.send(embed=embed, view=view)
        await ctx.message.delete()  # Delete the command message
        
    except Exception as e:
        await ctx.send(f"❌ Error setting up verification: {str(e)}")

@bot.command(name='stats')
@is_admin()
async def stats(ctx):
    """Show verification statistics"""
    try:
        async with aiohttp.ClientSession() as session:
            async with session.get(f"{API_BASE}?action=getVerifiedUsers") as response:
                if response.status == 200:
                    data = await response.json()
                    if data.get('success'):
                        users = data.get('users', [])
                        total_verified = len(users)
                        
                        # Count users from this server
                        server_verified = len([u for u in users if u.get('server_id') == str(ctx.guild.id)])
                        
                        embed = discord.Embed(
                            title="📊 Verification Statistics",
                            color=0x57F287
                        )
                        embed.add_field(name="Total Verified Users", value=f"{total_verified}", inline=True)
                        embed.add_field(name="From This Server", value=f"{server_verified}", inline=True)
                        embed.add_field(name="Server ID", value=f"{ctx.guild.id}", inline=True)
                        embed.set_footer(text=f"Requested by {ctx.author}")
                        
                        await ctx.send(embed=embed)
                    else:
                        await ctx.send(f"❌ Error fetching stats: {data.get('message', 'Unknown error')}")
                else:
                    await ctx.send(f"❌ Error fetching stats: HTTP {response.status}")
                    
    except Exception as e:
        await ctx.send(f"❌ Error fetching stats: {str(e)}")

@bot.command(name='pull')
@is_admin()
async def pull_users(ctx, limit: int = None):
    """Pull verified users to this server"""
    try:
        # Get verified users
        async with aiohttp.ClientSession() as session:
            async with session.get(f"{API_BASE}?action=getVerifiedUsers") as response:
                if response.status != 200:
                    await ctx.send(f"❌ Error fetching verified users: HTTP {response.status}")
                    return
                
                data = await response.json()
                if not data.get('success'):
                    await ctx.send(f"❌ Error fetching verified users: {data.get('message', 'Unknown error')}")
                    return
                
                users = data.get('users', [])
                if not users:
                    await ctx.send("❌ No verified users found")
                    return
                
                # Apply limit if specified
                if limit and limit > 0:
                    users = users[:limit]
                
                # Extract user IDs
                user_ids = [user['user_id'] for user in users]
                
                # Pull users using admin API
                pull_data = {
                    'guildId': str(ctx.guild.id),
                    'userIds': user_ids
                }
                
                async with session.post(
                    f"{API_BASE}?action=pullUsers",
                    json=pull_data,
                    headers={'Content-Type': 'application/json'}
                ) as pull_response:
                    if pull_response.status == 200:
                        result = await pull_response.json()
                        if result.get('success'):
                            embed = discord.Embed(
                                title="✅ Pull Operation Complete",
                                description=result.get('message', 'Users pulled successfully'),
                                color=0x57F287
                            )
                            
                            # Show results if available
                            if 'results' in result:
                                success_count = len([r for r in result['results'] if r.get('success')])
                                total_count = len(result['results'])
                                embed.add_field(
                                    name="Results",
                                    value=f"✅ {success_count}/{total_count} users added successfully",
                                    inline=False
                                )
                            
                            await ctx.send(embed=embed)
                        else:
                            await ctx.send(f"❌ Pull failed: {result.get('message', 'Unknown error')}")
                    else:
                        await ctx.send(f"❌ Pull failed: HTTP {pull_response.status}")
                        
    except Exception as e:
        await ctx.send(f"❌ Error pulling users: {str(e)}")

class VerificationView(discord.ui.View):
    def __init__(self, verify_url):
        super().__init__(timeout=None)
        self.verify_url = verify_url
        
        # Add verification button
        button = discord.ui.Button(
            label="Verify with Discord",
            style=discord.ButtonStyle.primary,
            emoji="🔐",
            url=verify_url
        )
        self.add_item(button)

@bot.event
async def on_command_error(ctx, error):
    if isinstance(error, commands.CheckFailure):
        await ctx.send("❌ You need administrator permissions to use this command!")
    elif isinstance(error, commands.CommandNotFound):
        pass  # Ignore unknown commands
    else:
        await ctx.send(f"❌ An error occurred: {str(error)}")
        print(f"Error in command {ctx.command}: {error}")

# Help command
@bot.command(name='help_verify')
async def help_verify(ctx):
    """Show help for verification commands"""
    embed = discord.Embed(
        title="🤖 Discord Verification Bot Help",
        description="Commands for server verification system",
        color=0x5865F2
    )
    
    embed.add_field(
        name="!setup_verify",
        value="Set up verification for your server (Admin only)",
        inline=False
    )
    embed.add_field(
        name="!stats",
        value="Show verification statistics (Admin only)",
        inline=False
    )
    embed.add_field(
        name="!pull [limit]",
        value="Pull verified users to this server (Admin only)\nOptional limit parameter to limit number of users",
        inline=False
    )
    embed.add_field(
        name="!help_verify",
        value="Show this help message",
        inline=False
    )
    
    embed.set_footer(text="All admin commands require Administrator permission")
    await ctx.send(embed=embed)

if __name__ == "__main__":
    if not BOT_TOKEN:
        print("❌ Error: DISCORD_BOT_TOKEN not found in environment variables")
        print("Please set your bot token in the .env file")
    else:
        try:
            bot.run(BOT_TOKEN)
        except discord.LoginFailure:
            print("❌ Error: Invalid bot token")
        except Exception as e:
            print(f"❌ Error starting bot: {e}")

File: bot/discord-bot.js
const { Client, GatewayIntentBits, PermissionFlagsBits, EmbedBuilder, ActionRowBuilder, ButtonBuilder, ButtonStyle } = require('discord.js');
const { createClient } = require('@blinkdotnew/sdk');

// Initialize Blink client
const blink = createClient({
  projectId: process.env.BLINK_PROJECT_ID,
  authRequired: false
});

const client = new Client({
  intents: [
    GatewayIntentBits.Guilds,
    GatewayIntentBits.GuildMessages,
    GatewayIntentBits.MessageContent,
    GatewayIntentBits.GuildMembers
  ]
});

const DISCORD_CLIENT_ID = '1397971356490006558';
const VERIFIED_ROLE_ID = process.env.DISCORD_VERIFIED_ROLE_ID;

client.once('ready', () => {
  console.log(`✅ Bot is ready! Logged in as ${client.user.tag}`);
});

client.on('messageCreate', async (message) => {
  if (message.author.bot) return;

  // Check if user has administrator permissions
  if (!message.member.permissions.has(PermissionFlagsBits.Administrator)) {
    return;
  }

  const args = message.content.split(' ');
  const command = args[0].toLowerCase();

  try {
    switch (command) {
      case '!setup_verify':
        await handleSetupVerify(message);
        break;
      case '!pull':
        await handlePull(message);
        break;
      case '!stats':
        await handleStats(message);
        break;
    }
  } catch (error) {
    console.error('Command error:', error);
    message.reply('❌ An error occurred while processing the command.');
  }
});

async function handleSetupVerify(message) {
  const serverId = message.guild.id;
  const serverName = message.guild.name;
  const setupBy = message.author.id;
  
  // Store server info in database
  await blink.db.botServers.create({
    id: `server_${serverId}`,
    serverId: serverId,
    serverName: serverName,
    setupBy: setupBy,
    verificationUrl: `https://discord-verification-bot-webpage-7tk0rize.sites.blink.new/verify?server=${serverId}`,
    isActive: 1
  });

  const embed = new EmbedBuilder()
    .setColor('#5865F2')
    .setTitle('🔐 Verification System Setup')
    .setDescription('Click the button below to get the verification link for your server!')
    .addFields(
      { name: '📋 Server', value: serverName, inline: true },
      { name: '👤 Setup by', value: `<@${setupBy}>`, inline: true }
    )
    .setTimestamp();

  const row = new ActionRowBuilder()
    .addComponents(
      new ButtonBuilder()
        .setLabel('🔗 Get Verification Link')
        .setStyle(ButtonStyle.Link)
        .setURL(`https://discord-verification-bot-webpage-7tk0rize.sites.blink.new/verify?server=${serverId}`)
    );

  await message.reply({ embeds: [embed], components: [row] });
}

async function handlePull(message) {
  const serverId = message.guild.id;
  
  // Get all verified users from database
  const verifiedUsers = await blink.db.verifiedUsers.list();
  
  if (verifiedUsers.length === 0) {
    return message.reply('❌ No verified users found.');
  }

  let successCount = 0;
  let failCount = 0;
  const results = [];

  for (const userData of verifiedUsers) {
    try {
      // Try to add user to the server
      const member = await message.guild.members.fetch(userData.userId).catch(() => null);
      
      if (member) {
        // User is already in the server, just add the role
        if (VERIFIED_ROLE_ID) {
          await member.roles.add(VERIFIED_ROLE_ID);
        }
        results.push(`✅ <@${userData.userId}> - Added role`);
        successCount++;
      } else {
        // User needs to be invited to the server
        results.push(`⚠️ <@${userData.userId}> - Not in server (needs invite)`);
        failCount++;
      }
    } catch (error) {
      console.error(`Error processing user ${userData.userId}:`, error);
      results.push(`❌ <@${userData.userId}> - Error`);
      failCount++;
    }
  }

  const embed = new EmbedBuilder()
    .setColor(successCount > failCount ? '#57F287' : '#ED4245')
    .setTitle('📥 Pull Results')
    .setDescription(`Processed ${verifiedUsers.length} verified users`)
    .addFields(
      { name: '✅ Success', value: successCount.toString(), inline: true },
      { name: '❌ Failed', value: failCount.toString(), inline: true },
      { name: '📊 Total', value: verifiedUsers.length.toString(), inline: true }
    )
    .setTimestamp();

  // Split results into chunks if too long
  const resultText = results.slice(0, 10).join('\n');
  if (resultText.length > 0) {
    embed.addFields({ name: '📋 Results (First 10)', value: resultText });
  }

  await message.reply({ embeds: [embed] });
}

async function handleStats(message) {
  try {
    // Get verified users count using the API
    const response = await fetch('https://7tk0rize--admin-api.functions.blink.new', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ action: 'stats' })
    });

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }

    const data = await response.json();
    
    if (!data.success) {
      throw new Error(data.error || 'Failed to fetch stats');
    }

    const totalVerified = data.total;
    
    // Get server count
    const servers = await blink.db.botServers.list();
    const totalServers = servers.length;
    
    // Get recent verifications (last 24 hours) - simplified for now
    const recentVerifications = 0; // Will be calculated properly later

    const embed = new EmbedBuilder()
      .setColor('#5865F2')
      .setTitle('📊 Verification Statistics')
      .setDescription('Current verification system stats')
      .addFields(
        { name: '👥 Total Verified Users', value: totalVerified.toString(), inline: true },
        { name: '🏠 Active Servers', value: totalServers.toString(), inline: true },
        { name: '🕐 Last 24h', value: recentVerifications.toString(), inline: true }
      )
      .setTimestamp()
      .setFooter({ text: `Requested by ${message.author.username}` });

    await message.reply({ embeds: [embed] });
  } catch (error) {
    console.error('❌ Error fetching stats:', error);
    message.reply('❌ Error fetching statistics.');
  }
}

// Handle user joining servers
client.on('guildMemberAdd', async (member) => {
  try {
    // Check if user is verified
    const verifiedUsers = await blink.db.verifiedUsers.list({
      where: { userId: member.user.id }
    });
    
    if (verifiedUsers.length > 0 && VERIFIED_ROLE_ID) {
      // Add verified role
      await member.roles.add(VERIFIED_ROLE_ID);
      console.log(`✅ Added verified role to ${member.user.tag}`);
    }
  } catch (error) {
    console.error('Error handling member join:', error);
  }
});

client.login(process.env.DISCORD_BOT_TOKEN);